File: backend/app/services/investor_profile.py
"""Investor-profile answer validation, deterministic scoring, and persistence."""

from datetime import UTC, datetime
from uuid import uuid4

from sqlalchemy.orm import Session

from app.core.errors import ServiceError
from app.domain.investor_profile.questionnaire import (
    QuestionnaireValidationError,
    resolve_investor_answers,
)
from app.domain.investor_profile.scoring import (
    SCORING_RULE_VERSION,
    SCORING_VERSION,
    score_investor_profile,
)
from app.models import InvestorProfileAssessment
from app.repositories import RecommendationRepository
from app.schemas.api import InvestorProfileAnalyzeRequest, InvestorProfileResponse


class InvestorProfileService:
    def __init__(
        self,
        session: Session,
    ) -> None:
        self.session = session
        self.repo = RecommendationRepository(session)

    @staticmethod
    def _response(assessment: InvestorProfileAssessment) -> InvestorProfileResponse:
        return InvestorProfileResponse(
            assessment_id=assessment.id,
            questionnaire_version=assessment.questionnaire_version,
            analysis_version=assessment.analysis_version,
            risk_score=assessment.risk_score,
            profile_type=assessment.profile_type,
            tendency_line=assessment.tendency_line,
            description=assessment.description,
            traits={
                "stability": assessment.stability,
                "return_seeking": assessment.return_seeking,
                "horizon": assessment.horizon,
            },
            analysis_summary=assessment.analysis_summary,
            model_version=assessment.model_version,
            created_at=assessment.created_at,
        )

    def _require_ai_consent(self, user_id: int) -> None:
        if not self.repo.has_ai_personalization_consent(user_id):
            raise ServiceError(
                "AI_PERSONALIZATION_CONSENT_REQUIRED",
                "AI 기반 맞춤형 서비스 이용 동의가 필요합니다.",
                403,
            )

    async def analyze(self, user_id: int, request: InvestorProfileAnalyzeRequest) -> InvestorProfileResponse:
        self._require_ai_consent(user_id)
        try:
            answers = resolve_investor_answers(
                request.questionnaire_version,
                ((item.question_id, item.option_id) for item in request.answers),
            )
        except QuestionnaireValidationError as exc:
            raise ServiceError(exc.code, str(exc), 400) from exc

        result = score_investor_profile(answers)
        assessment = InvestorProfileAssessment(
            id=uuid4(),
            user_id=user_id,
            questionnaire_version=request.questionnaire_version,
            analysis_version="v2",
            risk_score=result.risk_score,
            profile_type=result.profile_type,
            stability=result.stability,
            return_seeking=result.return_seeking,
            horizon=result.horizon,
            tendency_line=result.tendency_line,
            description=result.description,
            analysis_summary=result.analysis_summary,
            model_version=SCORING_VERSION,
            prompt_version=SCORING_RULE_VERSION,
            created_at=datetime.now(UTC),
        )
        try:
            self.session.add(assessment)
            self.session.commit()
        except Exception:
            self.session.rollback()
            raise
        return self._response(assessment)

    def latest(self, user_id: int) -> InvestorProfileResponse:
        assessment = self.repo.latest_assessment(user_id)
        if not assessment:
            raise ServiceError("INVESTOR_PROFILE_NOT_FOUND", "저장된 투자성향을 찾을 수 없습니다.", 404)
        return self._response(assessment)

File: frontend/src/lib/investorProfile.ts
import type { InvestorProfileResponse } from './backendApi';
import { RISK_QUESTIONS } from '../data/riskQuestions';

export interface InvestorTraits {
  /** 안정성 선호 — 1~5 (dot 채움 개수) */
  stability: number;
  /** 수익추구 성향 — 1~5 */
  returnSeeking: number;
  /** 투자기간 성향 — 1~5 */
  horizon: number;
}

export interface InvestorProfileResult {
  type: string;
  /** 0~100 최종 위험 점수. 점수 도입 전 v1 결과는 null이다. */
  riskScore: number | null;
  tendencyLine: string;
  description: string;
  traits: InvestorTraits;
}

/** 백엔드 POST /investor-profile/analyze · GET /investor-profile/me/latest 응답(snake_case traits)을
 *  화면이 실제로 쓰는 InvestorProfileResult 모양으로 맞춘다. */
export function mapInvestorProfileResponse(response: InvestorProfileResponse): InvestorProfileResult {
  return {
    type: response.profile_type,
    riskScore: response.risk_score ?? null,
    tendencyLine: response.tendency_line,
    description: response.description,
    traits: {
      stability: response.traits.stability,
      returnSeeking: response.traits.return_seeking,
      horizon: response.traits.horizon,
    },
  };
}

/** answers(문항별로 고른 보기 인덱스) → 백엔드 POST /investor-profile/analyze 가 기대하는
 *  { question_id, option_id } 배열로 변환한다. RISK_QUESTIONS 는 백엔드 설문 카탈로그
 *  (app/domain/investor_profile/questionnaire.py, QUESTIONNAIRE_V1)와 문항 순서·id 가 1:1로 맞춰져 있다. */
export function buildInvestorAnswerPayload(answers: number[]): { question_id: string; option_id: string }[] {
  return RISK_QUESTIONS.map((question, i) => ({
    question_id: question.id,
    option_id: question.options[answers[i]].id,
  }));
}

/** 결과·확인 화면에서 원문 보기를 다시 보여줄 때 쓰는 헬퍼 */
export function answerLabel(questionIndex: number, optionIndex: number | null | undefined): string {
  if (optionIndex == null) return '-';
  return RISK_QUESTIONS[questionIndex]?.options[optionIndex]?.title ?? '-';
}
